src/tests/unit/tests/common/message-creators/action-message-dispatcher.test.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { It, Mock, Times } from 'typemoq';
+import { Message } from '../../../../../common/message';
+import { ActionMessageDispatcher } from '../../../../../common/message-creators/action-message-dispatcher';
+import { Messages } from '../../../../../common/messages';
+import { BaseTelemetryData, TelemetryEventSource } from '../../../../../common/telemetry-events';
+
+describe('ActionMessageDispatcher', () => {
+    const postMessageMock = Mock.ofInstance((message: Message) => {});
+
+    const tabIds = [-1, null];
+
+    beforeEach(() => {
+        postMessageMock.reset();
+    });
+
+    describe('dispatchMessage', () => {
+        it.each(tabIds)('with tabId = %o', tabId => {
+            const message: Message = {
+                messageType: 'test-message-type',
+            };
+
+            const testObject = new ActionMessageDispatcher(postMessageMock.object, tabId);
+
+            testObject.dispatchMessage(message);
+
+            const expectedMessage = {
+                ...message,
+            };
+
+            if (tabId != null) {
```

Comment:
repeating the implementation in the test like this isn't great; it's disproportionately likely for a bug in this sort of logic to be in the choice of which comparator to use, for example. I think I'd prefer to see this sort of test *not* use `it.each`, and to have the individual tests write out explicitly the exact output they're expecting.

For example, consider how close the techinique of dynamically building up the message comes here to making the test accidentally depend on the way the implementation accidentally mutates the input message.

Reply:
`if` block has been removed now